### ISSUE TITLE
[Bug] NullReferenceException when invoking ITokenAcquisition.GetAccessTokenForAppAsync

### ISSUE THREAD

**Issue by gravity00** [AzureAD/microsoft-identity-web#1348]
**Which version of Microsoft Identity Web are you using?**
Microsoft Identity Web 1.15.1
Microsoft Identity Web MicrosoftGraph 1.15.1

**Where is the issue?**
* Web app
    * [ ] Sign-in users
    * [ ] Sign-in users and call web APIs
* Web API
    * [ ] Protected web APIs (validating tokens)
    * [ ] Protected web APIs (validating scopes)
    * [X] Protected web APIs call downstream web APIs
* Token cache serialization
     * [ ] In-memory caches
     * [ ] Session caches
     * [ ] Distributed caches
* Other (please describe)

**Is this a new or an existing app?**
This is a new app.

**Repro**

```csharp
using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authentication.JwtBearer;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc.Authorization;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Graph;
using Microsoft.Identity.Web;

namespace AzureActiveDirectoryTests
{
    public class Startup
    {
        private const string Instance = "https://login.microsoftonline.com/";
        private const string TenantId = "70d...";
        private const string Domain = "...";
        private const string ClientId = "276...";
        private const string ClientSecret = "ykLSY_...";

        public void ConfigureServices(IServiceCollection services)
        {
            services
                .AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
                .AddMicrosoftIdentityWebApi(o =>
                {
                    o.Events = new JwtBearerEvents
                    {
                        OnTokenValidated = OnTokenValidated
                    };
                }, o =>
                {
                    o.Instance = Instance;
                    o.TenantId = TenantId;
                    o.Domain = Domain;
                    o.ClientId = ClientId;
                    o.ClientSecret = ClientSecret;
                    o.AllowWebApiToBeAuthorizedByACL = true;
                })
                .EnableTokenAcquisitionToCallDownstreamApi(o =>
                {
                    o.Instance = Instance;
                    o.TenantId = TenantId;
                    o.ClientId = ClientId;
                    o.ClientSecret = ClientSecret;
                })
                .AddMicrosoftGraph(o =>
                {
                    o.BaseUrl = "https://graph.microsoft.com/v1.0";
                    o.Scopes = "User.Read Directory.Read.All";
                })
                .AddInMemoryTokenCaches();

            services.AddAuthorization(o =>
            {
                o.AddPolicy("RequireAuthenticatedUser", policy => policy
                    .AddAuthenticationSchemes(JwtBearerDefaults.AuthenticationScheme)
                    .RequireAuthenticatedUser()
                );
            });

            services
                .AddMvcCore(o =>
                {
                    o.Filters.Add(new AuthorizeFilter("RequireAuthenticatedUser"));
                })
                .AddApiExplorer();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment()) 
                app.UseDeveloperExceptionPage();

            app.UseRouting();

            app.UseAuthentication();
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });
        }

        private static async Task OnTokenValidated(TokenValidatedContext ctx)
        {
            var ct = ctx.HttpContext.RequestAborted;

            if (ctx.Principal?.Identity == null || ctx.SecurityToken == null) 
                return;

            var hasOverageOccurred = ctx.Principal.Claims.Any(c =>
                c.Type == "hasgroups" ||
                c.Type == "_claim_names" && c.Value == "{\"groups\":\"src1\"}"
            );
            if (!hasOverageOccurred) 
                return;

            var userId = ctx.Principal.Claims.SingleOrDefault(c =>
                c.Type == "http://schemas.microsoft.com/identity/claims/objectidentifier"
            )?.Value;

            if (string.IsNullOrWhiteSpace(userId)) 
                throw new InvalidOperationException("Object identifier claim was not found");

            var graphClient = ctx.HttpContext.RequestServices.GetRequiredService<GraphServiceClient>();

            // exception happens here
            var tmp = await graphClient.Users[userId]
                .MemberOf.Request()
                .Select("id,displayName")
                .GetAsync(ct);

            // code that would load groups as claims
        }
    }
}

```

**Expected behavior**
In cases were the flag `hasgroups` was set in the JWT during an implicit authentication, the API should be able to inquire the GraphAPI for the associated groups using a client credential flow. As far as we know, it is not possible to use the provided token in the request since the frontend must have a different `ClientId` than the backend (SPA invoking multiple APIs).

**Actual behavior**
An `NullReferenceException` is thrown when requesting the user member information.

```csharp
var tmp = await graphClient.Users[userId].MemberOf
    .Request()
    .Select(
        "id,displayName"
    )
    .GetAsync(ct);
```

Considering the stacktrace, it comes from an internal invocation to `MergedOptions.PrepareAuthorityInstanceForMsal()`.
Using directly the `ITokenAcquisition` also fails with the same exception:

```csharp
var tokenAcquisition = ctx.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>();

var graphClient = new GraphServiceClient(new DelegateAuthenticationProvider(async request =>
{
    // this will throw the NullReferenceException
    var token = await tokenAcquisition.GetAccessTokenForAppAsync(
        "https://graph.microsoft.com/.default",
        tokenAcquisitionOptions: new TokenAcquisitionOptions
        {
            CancellationToken = ct
        }
    );

    request.Headers.Authorization = new AuthenticationHeaderValue(
        "Bearer",
        token
    );
}));
```

Looking at the source code, seems the only possibility of this exception to happen is the `MergedOptions` has the property `Instance` as null (it looks like other properties aren't also set).

**Possible solution**
None, couldn't understand why all options have the property `Instance` set with a value, but internally in the `MergedOptions` it is null.

**Additional context / logs / screenshots**
```
1. Stacktrace when using the GraphServiceClient

 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Identity.Web.MergedOptions.PrepareAuthorityInstanceForMsal()
   at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForUserAsync(IEnumerable`1 scopes, String authenticationScheme, String tenantId, String userFlow, ClaimsPrincipal user, TokenAcquisitionOptions tokenAcquisitionOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetAccessTokenForUserAsync(IEnumerable`1 scopes, String authenticationScheme, String tenantId, String userFlow, ClaimsPrincipal user, TokenAcquisitionOptions tokenAcquisitionOptions)
   at Microsoft.Identity.Web.TokenAcquisitionAuthenticationProvider.AuthenticateRequestAsync(HttpRequestMessage request)
   at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.UserMemberOfCollectionWithReferencesRequest.GetAsync(CancellationToken cancellationToken)
   at AzureActiveDirectoryTests.Startup.OnTokenValidated(TokenValidatedContext ctx) in C:\Dev-Playground\AzureActiveDirectoryTests\AzureActiveDirectoryTests\Startup.cs:line

2. Stacktrace when using directly the ITokenAcquisition

---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Identity.Web.MergedOptions.PrepareAuthorityInstanceForMsal()
   at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForAppAsync(String scope, String authenticationScheme, String tenant, TokenAcquisitionOptions tokenAcquisitionOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetAccessTokenForAppAsync(String scope, String authenticationScheme, String tenant, TokenAcquisitionOptions tokenAcquisitionOptions)
   at AzureActiveDirectoryTests.TokenValidatedContextExtensions.<>c__DisplayClass0_0.<<EnsureGroupsAreLoadedAsync>b__1>d.MoveNext() in C:\Dev-Playground\AzureActiveDirectoryTests\AzureActiveDirectoryTests\TokenValidatedContextExtensions.cs:line 35
--- End of stack trace from previous location ---
   at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.UserMemberOfCollectionWithReferencesRequest.GetAsync(CancellationToken cancellationToken)
   at AzureActiveDirectoryTests.TokenValidatedContextExtensions.EnsureGroupsAreLoadedAsync(TokenValidatedContext ctx, CancellationToken ct) in C:\Dev-Playground\AzureActiveDirectoryTests\AzureActiveDirectoryTests\TokenValidatedContextExtensions.cs:line 51
```

![image](https://user-images.githubusercontent.com/3291285/127151029-63e41074-1514-456d-8d39-39fdda138477.png)

```csharp
        public void PrepareAuthorityInstanceForMsal()
        {
            if (IsB2C && Instance.EndsWith("/tfp/", StringComparison.OrdinalIgnoreCase))
            {
                Instance = Instance.Replace("/tfp/", string.Empty, StringComparison.OrdinalIgnoreCase).TrimEnd('/') + "/";
            }
            else
            {
                Instance = Instance.TrimEnd('/') + "/";
            }
        }
```

**Comment by jennyf19** [comment]
@gravity00 good catch. thanks for opening the issue.

**Comment by jennyf19** [comment]
included in [1.15.2 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.15.2)

**Comment by jennyf19** [comment]
@tsavela can you provide a link to your tests?

**Comment by tsavela** [comment]
Sure @jennyf19, here's a minimal repro: https://github.com/tsavela/microsoft-identity-web-bug-repro

**Comment by jennyf19** [comment]
@tsavela The mergeOptions are triggered from the JwtBearerOptions, usually from the `Authorize` attribute in a controller.
you can add the following:
```csharp
...
 var services = serviceCollection.BuildServiceProvider();
 
 var optionsMonitor = services.GetRequiredService<IOptionsMonitor<JwtBearerOptions>>();
 var options =  optionsMonitor.Get(JwtBearerDefaults.AuthenticationScheme);
 
 ...
```

and this `options.Get(JwtBearerDefaults.AuthenticationScheme)` will trigger the mergeOptions, so the instance is not null. we'll need to see how to handle this in Id Web, but this should unblock you for testing.

**Comment by jennyf19** [comment]
Included in [1.16.1 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.16.1)
